service/src/PlatformTopo.cpp

```diff
@@ -55,9 +63,44 @@ int geopm_read_cpuid(void)
     return ((family << 8) + model);
 }
 
+static volatile unsigned g_is_popen_complete = 0;
+static struct sigaction g_popen_complete_signal_action;
+
+static void geopm_topo_popen_complete(int signum)
+{
+    if (signum == SIGCHLD) {
+        g_is_popen_complete = 1;
+    }
+}
+
+int geopm_topo_popen(const char *cmd, FILE **fid)
+{
+    int err = 0;
+    *fid = NULL;
+
+    struct sigaction save_action;
+    g_popen_complete_signal_action.sa_handler = geopm_topo_popen_complete;
+    sigemptyset(&g_popen_complete_signal_action.sa_mask);
+    g_popen_complete_signal_action.sa_flags = 0;
+    err = sigaction(SIGCHLD, &g_popen_complete_signal_action, &save_action);
+    if (!err) {
+        *fid = popen(cmd, "r");
+        while (*fid && !g_is_popen_complete) {
+
+        }
+        g_is_popen_complete = 0;
+        sigaction(SIGCHLD, &save_action, NULL);
+    }
+    if (!err && *fid == NULL) {
+        err = errno ? errno : GEOPM_ERROR_RUNTIME;
+    }
```

Comment:
Minor suggestion: you can move this `if` into the previous `if` and reduce it to:

`if (*fid == NULL) {...`

since the previous `if` already ensures `!err` is true.